backend/python/workflows.py

```diff
@@ -0,0 +1,105 @@
+                retry_policy=self.retry_policy
+            )
+        except ApplicationError as e:
+            self.order_status.state = OrderStates.FAILED
+            message = "Failed to charge customer for " + product.name + ". Error: " + e.message
+            await workflow.execute_activity_method(
+                FoodDeliveryActivities.send_push_notification,
+                message,
+                start_to_close_timeout=timedelta(seconds=10),
+                retry_policy=self.retry_policy
+            )
+            raise ApplicationError(message, type=e.type)
+        self.order_status.state = OrderStates.PAID
+
+        try:
+            await workflow.wait_condition(
+                lambda: self.order_status.state == OrderStates.DELIVERED,
+                timeout=timedelta(seconds=30)
+            )
+        except asyncio.TimeoutError:
+            self.order_status.state = OrderStates.REFUNDING
+            await self.refund_and_notify(
+                product,
+                "⚠️  Your driver was unable to deliver your order. Your payment has been refunded."
+            )
+            raise ApplicationError("Not delivered in time", type="NotDeliveredInTime")
+        await workflow.execute_activity_method(
+            FoodDeliveryActivities.send_push_notification,
+            "✅  Order delivered!",
+            start_to_close_timeout=timedelta(seconds=10),
+            retry_policy=self.retry_policy
+        )
+
+        await asyncio.sleep(30)
+        await workflow.execute_activity_method(
+            FoodDeliveryActivities.send_push_notification,
+            "✍️  Rate your meal. How was the " + product.name + "?",
+            start_to_close_timeout=timedelta(seconds=10),
+            retry_policy=self.retry_policy
+        )
+
+        return "success"
+
+    @workflow.signal(name="pickedUp")
+    def picked_up_signal(self):
+        if self.order_status.state == OrderStates.PAID:
+            self.order_status.state = OrderStates.PICKED_UP
+
+    @workflow.signal(name="delivered")
+    def delivered_signal(self):
+        if self.order_status.state == OrderStates.PICKED_UP:
+            self.order_status.state = OrderStates.DELIVERED
+            self.order_status.delivered_at = workflow.now()
+
+    @workflow.query(name="getStatus")
+    def get_status_query(self) -> OrderStatus:
+        return self.order_status
+
+    async def refund_and_notify(self, product: Product, message: str):
+        await workflow.execute_activity_method(
+            FoodDeliveryActivities.refund_order,
+            product,
+            start_to_close_timeout=timedelta(seconds=10),
+            retry_policy=self.retry_policy
+        )
+        await workflow.execute_activity_method(
+            FoodDeliveryActivities.send_push_notification,
+            message,
+            start_to_close_timeout=timedelta(seconds=10),
+            retry_policy=self.retry_policy
+        )
```

Comment:
I would add a common send_notification method similar to refund_and_notify.  After you add the pickedUp logic, you will end up with many (3?) similar code blocks that send notifications.  You can simplify by adding a common method and calling it from the the 3 places.